File: 2015/MAC0327/Desafios 1/aula1/P5.c
/*******************************************************************************
 * Aluno:  André Meneghelli Vale,      Núm. USP: 4898948
 * Curso: Bacharelado em Ciências da Computação
 * Aula 1 - Desafio E
 * MAC0327 -- 08/07/2015 -- IME/USP, -- Prof. Cristina Gomes Fernandes
 * Compilador: gcc linux 4.8.4
 * Editor: Code Blocks;
 ******************************************************************************/

#include <stdio.h>
#include <stdlib.h>

struct actor
{
    unsigned int start;
    unsigned int end;
    unsigned int rooms;
} typedef Actor;

struct part
{
    unsigned int start;
    unsigned int end;
    int actor;
} typedef Part;

int backtrack(int part, int nparts, Part *parts, int nactors, Actor *actors)
{
    int i;

    if (part == nparts)
        return 1;

    for (i=0; i<nactors; i++)
    {
        if (parts[part].start<actors[i].start || parts[part].end>actors[i].end || actors[i].rooms==0)
            continue;
        actors[i].rooms--;
        parts[part].actor = i + 1;
        if (!backtrack(part+1, nparts, parts, 0, nactors, actors))
            actors[i].rooms++;
        else
            return 1;
    }
    return 0;
}

int main(int argc, const char* argv[])
{
    int nparts,
        nactors,
        i;
    Part *parts;
    Actor *actors;

    scanf("%d", &nparts);
    parts = (Part *) malloc(nparts*sizeof(Part));
    for (i=0; i<nparts; i++)
        scanf("%u %u", &parts[i].start, &parts[i].end);

    scanf("%d", &nactors);
    actors = (Actor *) malloc(nactors*sizeof(Actor));
    for (i=0; i<nactors; i++)
        scanf("%u %u %u", &actors[i].start, &actors[i].end, &actors[i].rooms);

    if (!backtrack(0, nparts, parts, nactors, actors)) printf("NO\n");
    else {
        printf("YES\n");
        for (i=0; i<nparts; i++)
            printf("%d ", parts[i].actor);
    }

    free(parts);
    free(actors);
    return 0;
}

File: 2015/MAC0327/Desafios 1/aula1/P5.py
# coding=utf-8
__author__ = 'andre'

"""
/*******************************************************************************
 * Aluno:  André Meneghelli Vale,      Núm. USP: 4898948
 * Curso: Bacharelado em Ciências da Computação
 * Aula 1 - Desafio E
 * MAC0327 -- 12/07/2015 -- IME/USP, -- Prof. Cristina Gomes Fernandes
 ******************************************************************************/
"""

def backtrack(index, parts):
    if index == len(parts):
        return True

    for actor in parts[index]['actors']:
        if actor['receiver'] == 0:
            continue
        actor['receiver'] -= 1
        parts[index]['actor'] = actor['index']
        if not backtrack(index+1, parts):
            actor['receiver'] += 1
            parts[index]['actor'] = False
        else:
            return True
    return False

def main():
    nparts = int(raw_input())
    parts = []
    for i in range(nparts):
        (start, end) = map(int, raw_input().split())
        parts.append({'index': i+1, 'start': start, 'end': end, 'actors': [], 'actor': False})
    nactors = int(raw_input())
    posibilities = 0
    actors = []
    for i in range(nactors):
        (start, end, receiver) = map(int, raw_input().split())
        actors.append({'index': i+1, 'start': start, 'end': end, 'receiver': receiver})
        posibilities += receiver
        for part in filter(lambda x: x['start'] >= actors[i]['start'] and x['end'] <= actors[i]['end'], parts):
            part['actors'].append(actors[i])

    parts.sort(key=lambda x: x['start'])
    for part in parts:
        part['actors'].sort(key=lambda x: x['end'])

    if posibilities < nparts or len(parts[0]['actors']) == 0:
        print "NO"
    else:
        if backtrack(0, parts):
            parts.sort(key=lambda x: x['index'])
            print "YES"
            print " ".join(map(lambda x: str(x['actor']), parts))
        else:
            print "NO"

if __name__ == "__main__":
    main()
